pulumitest/cacheStackState.go

```diff
@@ -0,0 +1,78 @@
+package pulumitest
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pulumi/providertest/pulumitest/opttest"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+)
+
+// CacheStackState caches the stack state from executing the `run` func. Subsequent calls will load the stack state from the cache.
+// If the cache does not exist, it will run the `run` func and cache the resulting stack state.
+// The `run` function is always executed in an isolated temp directory so it won't affect the original test.
+// The state exported from the `run` function (or cached from the previous run) is imported into the original test's stack.
+func (pulumiTest *PulumiTest) CacheStackState(cachePath string, run func(test *PulumiTest), opts ...opttest.Option) *PulumiTest {
```

Comment:
The name of the method feels a bit off since it not only executes the caching operation but is also used to simply read the state. Maybe something like `ReadOrInitStackState`?